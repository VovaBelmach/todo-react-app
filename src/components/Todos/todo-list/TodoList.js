import { useState } from "react";
import { useSelector, useDispatch } from "react-redux";
import {
  deleteTodo,
  setFilter,
  filteredTodosSelector,
  reorderTodos,
} from "../../../stores/todosSlice";
import TodoItem from "../todo-item/TodoItem";
import TodoFooter from "../todo-footer/TodoFooter";
import Card from "../../UI/Card/Card";

const TodoList = () => {
  const filteredTodos = useSelector(filteredTodosSelector);
  const [draggedIndex, setDraggedIndex] = useState(null);
  const dispatch = useDispatch();

  const onFilterHandler = (filterProperty) => {
    dispatch(setFilter(filterProperty));
  };

  const onDeleteCompletedHandler = () => {
    const completedTodoIds = filteredTodos
      .filter((todo) => todo.isCompleted)
      .map((todo) => todo.id);

    completedTodoIds.forEach((id) => {
      dispatch(
        deleteTodo({
          id: id,
        })
      );
    });
  };

  const onDragStartHandle = (event, index) => {
    setDraggedIndex(index);
    event.dataTransfer.setData("text/plain", index);
  };

  const onDragOverHandle = (event) => {
    event.preventDefault();
    event.dataTransfer.dropEffect = "move";
  };

  const onDropHandle = (event, index) => {
    event.preventDefault();
    dispatch(
      reorderTodos({
        draggedIndex: draggedIndex,
        droppedIndex: index,
      })
    );
  };

  const onDragEndHandle = () => {
    setDraggedIndex(null);
  };

  return (
    <Card>
      <ul>
        {filteredTodos.map((todo, index) => (
          <TodoItem
            key={todo.id}
            id={todo.id}
            index={index}
            isCompleted={todo.isCompleted}
            description={todo.description}
            onDragStartHandle={onDragStartHandle}
            onDragOverHandle={onDragOverHandle}
            onDropHandle={onDropHandle}
            onDragEndHandle={onDragEndHandle}
          />
        ))}
      </ul>
      <TodoFooter
        countTodos={filteredTodos.filter(todo => !todo.isCompleted).length}
        onFilterHandler={onFilterHandler}
        onDeleteCompletedHandler={onDeleteCompletedHandler}
      />
    </Card>
  );
};

export default TodoList;
