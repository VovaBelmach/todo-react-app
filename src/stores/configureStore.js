import { configureStore } from "@reduxjs/toolkit";
import todoReducer from "./todosSlice";
import { loadTodosStateFromLocalStorage } from "./helpers/todoLocalStorage.helper";
import { localStorageMiddleware } from "./middleware/localStorageRepositoryMiddleware"

export default configureStore({
  reducer: {
    todos: todoReducer,
  },
  preloadedState: loadTodosStateFromLocalStorage(),
  middleware: (getDefaultMiddleware) =>
    getDefaultMiddleware().concat(localStorageMiddleware),
});
