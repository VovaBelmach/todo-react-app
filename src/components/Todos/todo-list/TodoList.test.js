import React from "react";
import { render, screen } from "@testing-library/react";
import { Provider } from "react-redux";
import configureStore from "redux-mock-store";
import TodoList from "./TodoList";
import {
  TODO_LIST_EMPTY_LIST_MESSAGE,
  TODO_ITEM_DELETE_BUTTON_NAME,
  TODO_FILTER_ALL_BUTTON_NAME,
} from "../../../constants";

const mockStore = configureStore([]);

describe("TodoList component", () => {
  let store;

  beforeEach(() => {
    store = mockStore({
      todos: {
        todos: [
          {
            id: 'a74ba32f-3868-48f9-bb1a-2f8267844fc1',
            description: "Learn React",
            isCompleted: false,
          },
          {
            id: 'c4eb6856-96af-41ef-b497-7424120cbbc4',
            description: "Write unit tests",
            isCompleted: true,
          },
        ],
        filter: TODO_FILTER_ALL_BUTTON_NAME,
      },
    });
  });

  it("should render correctly when there are no filtered todos", () => {
    // Arrange
    store = mockStore({
      todos: {
        todos: [],
        filter: TODO_FILTER_ALL_BUTTON_NAME,
      },
    });
    
    const { container } = render(
      <Provider store={store}>
        <TodoList />
      </Provider>
    );

    // Act
    expect(container).toMatchSnapshot();
  });

  it("should render correctly when there are filtered todos", () => {
    // Arrange
    const { container } = render(
      <Provider store={store}>
        <TodoList />
      </Provider>
    );

    // Act
    expect(container).toMatchSnapshot();
  });

  it("should render a list of todo items", () => {
    // Arrange
    render(
      <Provider store={store}>
        <TodoList />
      </Provider>
    );

    // Assert
    const todoItems = screen.getAllByRole("todo-item");
    expect(todoItems.length).toBe(2);
  });

  it("should render a message when no todos are present", () => {
    // Arrange
    store = mockStore({
      todos: {
        todos: [],
        filter: TODO_FILTER_ALL_BUTTON_NAME,
      },
    });

    render(
      <Provider store={store}>
        <TodoList />
      </Provider>
    );

    // Assert
    const message = screen.getByText(TODO_LIST_EMPTY_LIST_MESSAGE);
    expect(message).toBeInTheDocument();
  });

  it("dispatches deleteTodo action when delete completed button is clicked", () => {
    // Arrange
    render(
      <Provider store={store}>
        <TodoList />
      </Provider>
    );

    // Act
    const deleteButton = screen.getAllByRole("button", {
      name: TODO_ITEM_DELETE_BUTTON_NAME,
    });
    deleteButton[0].click();

    // Assert
    expect(store.getActions()).toEqual([
      {
        type: "todos/deleteTodo",
        payload: {
          id: 'a74ba32f-3868-48f9-bb1a-2f8267844fc1',
        },
      },
    ]);
  });
});
