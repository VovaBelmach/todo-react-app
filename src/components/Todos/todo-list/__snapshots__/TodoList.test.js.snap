// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList component should render correctly when there are filtered todos 1`] = `
<div>
  <div
    class="card todo-list"
  >
    <ul>
      <li
        class="incomplete"
        draggable="true"
        role="todo-item"
        style="opacity: 1;"
      >
        <input
          class="checkbox"
          id="a74ba32f-3868-48f9-bb1a-2f8267844fc1"
          name="a74ba32f-3868-48f9-bb1a-2f8267844fc1"
          type="checkbox"
        />
        <label
          for="a74ba32f-3868-48f9-bb1a-2f8267844fc1"
        >
          Learn React
        </label>
        <button
          class="button delete-btn"
          type="button"
        >
          X
        </button>
      </li>
      <li
        class="complete"
        draggable="true"
        role="todo-item"
        style="opacity: 1;"
      >
        <input
          checked=""
          class="checkbox"
          id="c4eb6856-96af-41ef-b497-7424120cbbc4"
          name="c4eb6856-96af-41ef-b497-7424120cbbc4"
          type="checkbox"
        />
        <label
          for="c4eb6856-96af-41ef-b497-7424120cbbc4"
        >
          Write unit tests
        </label>
        <button
          class="button delete-btn"
          type="button"
        >
          X
        </button>
      </li>
    </ul>
    <footer
      class="todo-footer"
    >
      <span>
        1 todo left
      </span>
      <ul
        class="todo-filter"
      >
        <li>
          <button
            class="button "
            type="button"
            value="All"
          >
            All
          </button>
        </li>
        <li>
          <button
            class="button "
            type="button"
            value="Active"
          >
            Active
          </button>
        </li>
        <li>
          <button
            class="button "
            type="button"
            value="Completed"
          >
            Completed
          </button>
        </li>
      </ul>
      <button
        class="button todo-clear-completed"
        type="button"
        value="Clear completed"
      >
        Clear completed
      </button>
    </footer>
  </div>
</div>
`;

exports[`TodoList component should render correctly when there are no filtered todos 1`] = `
<div>
  <div
    class="card todo-list"
  >
    <p>
      All todos completed! Add a todo and it will be displayed here.
    </p>
    <footer
      class="todo-footer"
    >
      <span>
        All done!
      </span>
      <ul
        class="todo-filter"
      >
        <li>
          <button
            class="button "
            type="button"
            value="All"
          >
            All
          </button>
        </li>
        <li>
          <button
            class="button "
            type="button"
            value="Active"
          >
            Active
          </button>
        </li>
        <li>
          <button
            class="button "
            type="button"
            value="Completed"
          >
            Completed
          </button>
        </li>
      </ul>
      <button
        class="button todo-clear-completed"
        type="button"
        value="Clear completed"
      >
        Clear completed
      </button>
    </footer>
  </div>
</div>
`;
